tidy3d/components/types.py

```diff
@@ -26,71 +24,76 @@ def annotate_type(UnionType):  # pylint:disable=invalid-name
-            if isinstance(v, list)
-            else cls.inner_type(v)  # pylint:disable=no-member
-        )
+    ndim = None
+    dtype = None
 
     @classmethod
     def __get_validators__(cls):
-        """boilerplate"""
-        yield cls.validate_type
+        yield cls.load_complex
+        yield cls.convert_to_numpy
+        yield cls.check_dims
 
     @classmethod
-    def validate_type(cls, val):
-        """validator"""
-        # need to fix, doesnt work for simulationdata_export and load?
+    def load_complex(cls, val):
+        """Special handling to load a complex-valued np.ndarray saved to file."""
+        if not isinstance(val, dict):
+            return val
+        if "real" not in val or "imag" not in val:
+            raise ValueError("ArrayLike real and imaginary parts not stored properly.")
+        arr_real = np.array(val["real"])
+        arr_imag = np.array(val["imag"])
+        return arr_real + 1j * arr_imag
 
-        if isinstance(val, np.ndarray):
-            val_ndims = len(val.shape)
-            cls_ndims = cls.ndims  # pylint:disable=no-member
-
-            if (cls_ndims is not None) and (cls_ndims != val_ndims):
-                raise ValidationError(
-                    "wrong number of dimensions given. " f"Given {val_ndims}, expected {cls_ndims}."
-                )
-            return cls.make_tuple(val.tolist())
+    @classmethod
+    def convert_to_numpy(cls, val):
+        """Convert the value to np.ndarray and provide some casting."""
+        return np.array(val, ndmin=1, dtype=cls.dtype, copy=True)
 
-        return tuple(val)
+    @classmethod
+    def check_dims(cls, val):
+        """Make sure the number of dimensions is correct."""
+        if cls.ndim and val.ndim != cls.ndim:
+            raise ValidationError(f"Expected {cls.ndim} dimensions for ArrayLike, got {val.ndim}.")
+        return val
 
     @classmethod
     def __modify_schema__(cls, field_schema):
-        """Sets the schema of ArrayLike."""
-
-        field_schema.update(
-            dict(
-                title="Array Like",
-                description="Accepts sequence (tuple, list, numpy array) and converts to tuple.",
-                type="tuple",
-                properties={},
-                required=[],
-            )
-        )
-
+        """Sets the schema of DataArray object."""
 
-class ArrayLikeMeta(type):
-    """metclass for Array, enables Array[type] -> TypedArray"""
+        schema = dict(
+            title="ArrayLike",
+            type="ArrayLike",
+        )
+        field_schema.update(schema)
 
-    def __getitem__(cls, type_ndims):
-        """Array[type, ndims] -> TypedArrayLike"""
-        desired_type, ndims = type_ndims
-        return type("Array", (TypedArrayLike,), {"inner_type": desired_type, "ndims": ndims})
 
+def constrained_array(dtype: type = None, ndim: int = None) -> type:
+    """Generate an ArrayLike sub-type with constraints built in."""
+    type_name = "ArrayLike_"
```

Comment:
Doesn't it make more sense to add the separator in the if bodies?
```python
type_name = "ArrayLike"
if dtype is not None:
    type_name += f"_dtype={dtype}"
if ndim is not None:
    type_name += f"_ndim={ndim}"
```

Reply:
yea you're right. I'll make the change.